core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -0,0 +1,413 @@
+                ))
+                lookForMissing()
+              } else {
+                log.debug("Updating offset to {} from pId {} seqNr {}", currentOffset, pid, repr.persistentRepr.sequenceNr)
+                currentOffset = row.getUUID("timestamp")
+                tagPidSequenceNrs += (repr.persistentRepr.persistenceId -> ((expectedSequenceNr, repr.offset)))
+                push(out, repr)
+                tryPushOne()
+              }
+            }
+
+          case QueryResult(rs) =>
+            if (rs.isExhausted) {
+              queryExhausted()
+            } else if (rs.getAvailableWithoutFetching == 0) {
+              log.debug("Fully fetched, getting more for next pull (not implemented yet)")
+            }
+
+          case BufferedEvents(events) if isAvailable(out) =>
+            log.debug("Pushing buffered event")
+            events match {
+              case Nil =>
+                continue()
+              case e :: Nil =>
+                push(out, e)
+                queryState = QueryIdle
+                query()
+              case e :: tail =>
+                push(out, e)
+                queryState = BufferedEvents(tail)
+                tryPushOne()
+            }
+
+          case _ =>
+            log.debug("Trying to push one but no query results currently {}", queryState)
+        }
+      }
+
+      private def queryExhausted(): Unit = {
+        log.debug("Query exhausted, rows for this query: {}. New time uuid {}", rowsForCurrentQuery, currentOffset)
+        if (missingLookup.isDefined) {
+          val missing = missingLookup.get
+          if (missing.queryPrevious) {
+            log.info("{}: Missing could be in previous bucket. Querying right away.", session.tag)
+            missingLookup = missingLookup.map(m => m.copy(
+              queryPrevious = false,
+              bucket = m.bucket.previous(1)
+            ))
+            lookForMissing()
+          } else {
+            log.info("{}: Still looking for missing. {}. Will wait for next poll.", session.tag, missingLookup)
+            queryState = QueryIdle
+            missingLookup = missingLookup.map(m => {
+              val newBucket = TimeBucket(m.maxOffset, bucketSize)
+              m.copy(
+                bucket = newBucket,
+                queryPrevious = !newBucket.within(m.previousOffset)
+              )
+            })
+            // a current query doesn't have a poll we schedule
+            if (refreshInterval.isEmpty)
+              scheduleOnce(QueryPoll, settings.refreshInterval)
+          }
+        } else if (currTimeBucket.inPast) {
+          nextTimeBucket()
+          log.debug("Moving to next bucket: {}", currTimeBucket)
+          query()
+        } else {
+          if (toOffset.isDefined) {
+            log.debug("At today, ending.")
+            completeStage()
+          } else {
+            queryState = QueryIdle
+            log.debug("Nothing todo, waiting for next poll")
```

Comment:
Yes if we exhaust the query and we're in the current bucket for live mode we go Idle and wait for the next poll. For current we complete the statge. Could make mode a field for readability rather than checking the refreshInterval?